src/govuk/components/hide-this-page/hide-this-page.js

```diff
@@ -0,0 +1,44 @@
+import { nodeListForEach } from '../../common'
+
+function HideThisPage ($module) {
+  this.$module = $module
+}
+
+HideThisPage.prototype.openNewTab = function (e) {
+  var isClickEvent = typeof e !== 'undefined'
+
+  if (isClickEvent) {
+    e.preventDefault()
+  }
+
+  var target = isClickEvent ? e.target : document.querySelector('.govuk-js-hide-this-page-button')
+
+  window.open(target.getAttribute('data-new-tab-url'), '_newtab')
+
+  // Check if the user's browser supports the history API
+  // If so, we apply a history scrambling enhancement
+  if (typeof window.history !== 'undefined' && typeof window.history === 'object') {
+    document.title = target.getAttribute('data-fake-page-title')
+    // Do not remove the empty value in the below function call. They are required to ensure we comply with the history API
+    // https://developer.mozilla.org/en-US/docs/Web/API/History_API
+    window.history.replaceState(target.getAttribute('data-fake-page-title'), '')
+  }
+
+  window.location.href = target.href
+}
+
+HideThisPage.prototype.init = function () {
+  // We put the loop here instead of in all.js because we want to have both listeners on the individual buttons for clicks
+  // and a single listener for the keyboard shortcut
+  nodeListForEach(this.$module, function ($button) {
+    $button.querySelector('.govuk-js-hide-this-page-button').addEventListener('click', this.openNewTab)
+  }.bind(this))
+
+  window.addEventListener('keydown', function (e) {
```

Comment:
There are potentially some edge cases where we don't want this escape key functionality to fire, in particular, with popover elements, such as those that appear when certain form elements are in use—datepickers, select lists and the like. These could lead to users inadvertently exiting a form and losing data or progress. 

I did a quick whip around and, luckily, browser vendors seem to mostly handle that for us. (✅ = event did not propagate to `window`; ❌ = event _did_ propagate to `window`). 

|Browser|`input[type="date"]` datepicker|`select` menu|
|:-|:-|:-|
|Chrome 99|✅|✅|
|Edge 99|✅|✅|
|Opera 84|✅|✅|
|Firefox 98|✅|✅|
|Safari 15.4 (macOS)|❌|✅|
|Safari 15.4 (iOS/iPadOS)|✅|✅|
|IE 11|N/A|✅|

It's also notable that both the accessible-autocomplete and the country-and-territory-autocomplete packages currently use the Escape key to close an open popover menu, and neither prevent propagation of that event to the `window` at this time. 

Given we don't encourage the use of `input[type="date"]` anyway, the Safari issue is perhaps not the highest priority at this time. I don't know if we want to simply document these situations or attempt to resolve these in code.

Reply:
This is incredibly thorough! Many thanks, I'll link this to the [keyboard shortcut issue](https://github.com/alphagov/govuk-design-system/issues/2116).

We're very aware of the risk of using `esc` and how it may conflict with a number of existing functionalities, however we're implementing it for now for the purpose of having an established convention for this pattern so we an test if keyboard shortcuts are actually useful or not. See [my comment on the keyboard issue](https://github.com/alphagov/govuk-design-system/issues/2116#issuecomment-1077837908) for further details.